Restrict prop graph topology interface

amberhassaan opened the pull request:
Move away from arrow arrays. Restrict access to internals of topology.
Create new graph instead of SetTopology

IMPORTANT: Wait for https://github.com/KatanaGraph/katana-enterprise/pull/1427 before mergin.

amberhassaan committed:
Rebasing squashed commits on top of master
Squashing all commits before a conflicting rebase on master

WIP: Remove arrow array views of graph topology

WIP: checkpoint: fix build errors

WIP: louvain_clustering cleanup after topology interface changes

Build fixes related to PG changes

WIP: GraphTopologyAccess Provide access to private members of GraphTopology

fix build. Fix silly r-value return type mistake

fix more of r-value return type mistakes

another r-value return error

bug fix: check if property exists before trying to remove it. Made
apparent by passing empty graph

r-value related build fixes

Missed return

Partial fix for changes to GraphTopology

python: Fix wrappers affected by changes to graph topo types.

stylistic changes

Remove GraphTopologyAccess. Address other comments

Fix build after rebase master

Lint error

Attempt at fixing Python build error

Bad fix. Fix again

User edges().end() instead of edge_end()

added noexcept

more noexcept. Remove resolved TODO

Bug Fix: Was using the wrong comparator for sorting dest array. Compiler
didn't catch it because of silent conversion  uint32_t -> uint64_t

Typo: build fix

added std::fill(). Removed enalbe_if from iota()

helper function to generate uinform random sequences

ddn0 commented on libgalois/include/katana/ParallelSTL.h:
```diff
@@ -419,6 +419,28 @@ transform(
+    typename T = typename std::iterator_traits<ForwardIt>::value_type,
+    typename __Unused = std::enable_if_t<std::is_arithmetic<T>::value>>
+void
+iota(const ForwardIt& first, const ForwardIt& last, const T& init_val) {
```
can we call `init_val`, `start_val` and/or add some documentation on this function explaining what it does in terms of its arguments?

amberhassaan replied:
Fixed: ec837c8a1326e3e22795c6e3bf7a43627f6a0233

ddn0 commented on libgalois/include/katana/ParallelSTL.h:
```diff
@@ -419,6 +419,28 @@ transform(
+template <
+    typename ForwardIt,
+    typename T = typename std::iterator_traits<ForwardIt>::value_type,
+    typename __Unused = std::enable_if_t<std::is_arithmetic<T>::value>>
```
nit: instead of using an arbitrary name for this type parameter can you use this pattern for SFINAE?

```cpp
std::enable_if_t<std::is_arithmetic_v<T>, void>
iota(const ForwardIt& first, ....) { ... }
```

putting it on the return value is the least likely to conflict with any current or future uses.

amberhassaan replied:
Fixed: ec837c8a1326e3e22795c6e3bf7a43627f6a0233

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -18,6 +18,21 @@
+template <
+    typename T,
+    typename __Unused = std::enable_if_t<std::is_arithmetic<T>::value>>
+auto
```
nit: I'd prefer an explicit return type and SFINAE on the return type instead of on a default value for a template parameter.

Also, identifiers starting with double underscores are generally reserved: https://en.cppreference.com/w/cpp/language/identifiers and we should avoid introducing them ourselves.

amberhassaan replied:
Chose return type pattern of enable_if

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -18,6 +18,21 @@
+
+/// Provides write access to private data of GraphTopology.
+/// WARNING: Dangerous
+class KATANA_EXPORT GraphTopologyAccess;
```
Shouldn't `KATANA_EXPORT` be on the class definition (as well)?

Also you can consider not exporting the access class, e.g.,

```cpp

class GraphTopologyAccess;

... friend class GraphTopologyAccess ...

class KATANA_EXPORT MyAccess: public GraphTopologyAccess {

};

```

amberhassaan replied:
Can you please explain the benefit? 
Related to this, I had a discussion with Roshan and Simon and this class may not be needed, so I might get rid of it.

amberhassaan replied:
Removed GraphTopologyAccess in 6f4b8a0971e97a7fcad213f68d6e2d97359f777c

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -41,49 +56,53 @@ class KATANA_EXPORT GraphTopology {
+    if (this == &that) {
+      return true;
+    }
+    if (this->num_nodes() != that.num_nodes()) {
```
nit: some preference to omit `this` here as we use `num_nodes()`  (without this) everywhere else in this class

amberhassaan replied:
Fixed: 6f4b8a0971e97a7fcad213f68d6e2d97359f777c

ddn0 commented on libgalois/include/katana/analytics/ClusteringImplementationBase.h:
```diff
@@ -442,61 +442,27 @@ struct ClusteringImplementationBase {
-      katana::PropertyGraph* pfg_from, katana::PropertyGraph* pfg_to,
-      const std::string& edge_property_name,
+  template <typename NodeData>
+  katana::Result<std::unique_ptr<katana::PropertyGraph>> CreateDuplicateGraph(
```
nit: why wouldn't we just call this function "DuplicateGraph"?

also, if this takes a template parameter, wouldn't it make more sense to return a TypedPropertyGraph instead?

amberhassaan replied:
Agree with renaming the function. Not sure if returning a TypedPropertyGraph is worth the change in the code. The call to Duplicate is inside an if-else, each of which produce a PropertyGraph pointer. While typed property graph is constructed afterwards in an iterative loop

ddn0 commented on libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp:
```diff
@@ -362,7 +362,7 @@ struct LouvainClusteringImplementation
      * computation proceeds.
      */
-    auto pfg_mutable = std::make_unique<katana::PropertyGraph>();
+    /*
```
can you just delete this code?

amberhassaan replied:
Done

ddn0 commented on libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp:
```diff
@@ -380,13 +380,17 @@ struct LouvainClusteringImplementation
     }
+    */
+
+    std::vector<std::string> temp_edge_property_names = {
```
There is: https://github.com/KatanaGraph/katana/blob/master/libgalois/include/katana/analytics/Utils.h#L75

amberhassaan replied:
Done.

ddn0 commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -1157,6 +1157,6 @@ katana::ConvertMongoDB(
     KATANA_LOG_FATAL("Failed to construct graph: {}", r.error());
   } else {
-    return r.value();
+    return std::move(r.value());
```
I'm surprised you don't get a warning about an unnecessary move here.

amberhassaan replied:
Because r.value() does not return the r-value reference. Compiler throws an error without std::move()

roshandathathri commented on libgalois/include/katana/LargeArray.h:
```diff
@@ -206,6 +206,28 @@ class LargeArray {
+      return true;
+    }
+
+    for (size_t i = 0, sz = left.size(); i < sz; ++i) {
```
Isn't this an expensive operation? I think it should be documented.

amberhassaan replied:
Done

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -41,49 +56,53 @@ class KATANA_EXPORT GraphTopology {
-  GraphTopology& operator = (GraphTopology&& that) {
+      : adj_indices_(std::move(adj_indices)), dests_(std::move(dests)) {}
+  // out_indices(LargeToArrowArray(adj_indices_)),
+  // out_dests(LargeToArrowArray(dests_)) {}
```
Drop

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -41,49 +56,53 @@ class KATANA_EXPORT GraphTopology {
+      return false;
+    }
+
+    return this->adj_indices_ == that.adj_indices_ &&
```
Ditto about documentation

amberhassaan replied:
Done

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -41,49 +56,53 @@ class KATANA_EXPORT GraphTopology {
-           out_dests->Equals(*other.out_dests);
+  const Edge* adj_data() const noexcept { return &adj_indices_[0]; }
+
+  const Node* dest_data() const noexcept { return &dests_[0]; }
```
Didn't like these low-level functions. Why not provide `const` getters for the two variables instead? The user can then get its data as a pointer.

amberhassaan replied:
I disagree with providing silly-old getters as they 1) expose internal type of the array 2) are prone to people (inadvertently) trying to copy, or worse, move the arrays.

amberhassaan replied:
I'd be happier if we introduce a Buffer type and return a buffer with `const T*` and `a size`, so the user doesn't rely on `num_nodes()` or `num_edges`

arthurp replied:
@amberhassaan you just described an arrow array pretty much. We could just wrap the underlying memory in an arrow array (that does not take ownership of the buffer) and return that.

arthurp replied:
That said, not providing any access to the buffers would be better. Best to just provide abstract accessors so no one can observe the internal data layout.

roshandathathri replied:
I agree with Arthur. We should either drop these or return them as an `arrow::Array`

amberhassaan replied:
Yes, but Arrow Arrays come with minor annoyances that add up to anger. Size is int64_t. Don't have `operator[]`. Not sure but seems like I need a `shared_ptr<arrow::Buffer>` to get started with creating the array. Feels too much. I just want a 
```
template <typename T>
Buffer {
  const T* data_;
  size_t size_;
...
};

// and may be
template <typename T>
MutableBuffer {
   T* data_;
   size_t size_;
};
```

arthurp replied:
Can we just remove these accessors then? They totally break the abstraction you were trying to create anyway.

amberhassaan replied:
Not really. I need a back-door. I can either bring back the GraphTopologyAccess class or I can use these accessors and do a `const_cast` to make it obvious that I'm doing something bad.

danielmawhirter replied:
[This RFC](https://docs.google.com/document/d/16XxEBlbaP7yBSxFPIB_XDRP4CyqxAfSdScR0Q12TFKU/edit?usp=sharing) contains my idea for how to wrap nodes and edges into arrow in the long term. It doesn't address the allocators, only the semantics, to allow us to continue using some of arrow's functionality w.r.t parquet, ipc, compute, etc. The custom classes created there could implement `operator[]`, and/or modifiability, which are not valid in arrow in general due to nullability.
This PR need not consider that if you don't want it to, but there seems to be enough of a discussion on semantics here to merit bringing it up.

arthurp replied:
@amberhassaan Where do you need the back door? Generally needing a back door implies the rest of the interface is incomplete. Maybe there needs to be a more capable constructor or Make function in this class.

amberhassaan replied:
@arthurp : I need it temporarily. There are functions like `SortEdgesByDest(pg_ptr)` that modify the topology arrays. Then some code in Partitioning (libgluon/Partition.cpp) that modifies the arrays. 
I share the desire of a future without such back doors.

amberhassaan replied:
@danielmawhirter : I agree that arrow Arrays offer benefit when accessing storage. I'm not convinced yet that they offer a benefit when it comes to in-memory topologies, which I'm hoping we'll have a few more of in the near future. 

Once storage makes use of arrow arrays for topology, it may make sense for in-memory representation to follow in vein, but at least for now, LargeArray (NumaArray) offers numa-aware performance benefits and works well with various places in code, such as computing transposed or sorted topologies, even in some analytics apps, and in libgluon. Most of these places were converting LargeArrays to Arrow Arrays and leaking memory. 

Also, once we figure out num-aware allocators for Arrow Arrays, we should revisit the topology representation.

arthurp replied:
I didn't realize it's temporary. Just leave them as returning pointers and add todos at their declarations and use. Hopefully we can update the other code to use the abstraction soon.

Re: NUMA Aware arrow arrays. I actually know how to do it. I just haven't spent the time. I assumed there was someone else tracking it. See https://arrow.apache.org/docs/cpp/memory.html#memory-pools. It's actually pretty straight forward. I don't think it matters that the allocator uses parallel sections. It's still just an allocator. Obviously we would need to use a different allocator for allocations inside parallel sections than in the main thread. But this is already true, IIUC.

roshandathathri replied:
@danielmawhirter is aware of these differences and is looking into it. We will need different types: `arrow::Array`, `katana::gstl::ArrowArray`, and `katana::NUMAArrowArray` (analogous to `std::vector`, `katana::gstl::Vector`, and `katana::NUMAArray` respectively).

arthurp replied:
I'll step back from this, because I have no understanding why a different class is needed for each of these cases. It seems like they should all the same interface to me. But it's not my area, so I'll leave it to y'all. I just worry that a proliferation of classes which provide extremely similar interfaces, but incompatibly, will cause maintenance problems in the future.

roshandathathri replied:
For arrow, we will NOT need different classes; just different types. The difference b/w the types will be different allocators. But all of them will NOT provide the same semantics. While `katana::gstl::Vector` can replace all instances of `std::vector`, `katana::NUMAArray` cannot (it's not about performance but about correctness).

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -98,15 +117,18 @@ class KATANA_EXPORT GraphTopology {
     return dests_[edge_id];
   }
 
+  Node edge_dest(const edge_iterator ei) const noexcept {
```
If this can be a `noexcept`, so can the function it calls, right?

amberhassaan replied:
Yes, they all should be noexcept :)

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -122,30 +144,33 @@ class KATANA_EXPORT GraphTopology {
-        arrow::MutableBuffer::Wrap(lg_arr.data(), lg_arr.size()));
-  }
+  Edge* adj_data() noexcept { return &adj_indices_[0]; }
+  Node* dest_data() noexcept { return &dests_[0]; }
```
Ditto about low-level functions. Provide direct access to the member variables and getters is better in my opinion. This seems too specific.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -122,30 +144,33 @@ class KATANA_EXPORT GraphTopology {
+
+  GraphTopology::Edge* adj_data() noexcept { return topo_->adj_data(); }
+
+  GraphTopology::Node* dest_data() noexcept { return topo_->dest_data(); }
```
This is fine.

roshandathathri commented on libgalois/src/BuildGraph.cpp:
```diff
@@ -1866,18 +1832,51 @@ ImportData::ValueFromArrowScalar(std::shared_ptr<arrow::Scalar> scalar) {
+  std::unique_ptr<katana::PropertyGraph> graph = std::move(pg_result.value());
+
+  if (graph_comps.nodes.properties->num_columns() > 0) {
+    auto result = graph->AddNodeProperties(graph_comps.nodes.properties);
```
@tylershunt Can the properties be moved?

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -116,21 +116,19 @@ WriteTopology(const katana::GraphTopology& topology) {
     static_assert(std::is_same_v<std::decay_t<decltype(*raw)>, uint32_t>);
-    auto buf = std::make_shared<arrow::Buffer>(
-        reinterpret_cast<const uint8_t*>(raw), num_edges * sizeof(uint32_t));
+    auto buf = arrow::Buffer::Wrap(raw, num_edges);
```
nit: I'd prefer a similar name instead of `ProjectAsArrowArray`. Maybe `WrapAsArrowArray`?

roshandathathri reviewed (approved):
LGTM

arthurp commented:
@amberhassaan I pushed updates to the python wrapper. The python tests are failing though. It's possible this is a bug in the wrapper, but there is also a crash in a call to `SortAllEdgesByDest` from the k-truss implementation, so I think the issues might be in the C++ implementation.

What I'm seeing are:

* Weird results of algorithms that don't match previous results. Note that the output checks are pretty strict, so if node order changed without actually affecting correctness that would explain some, *but not all* of the failures.
* Inconsistent output in the permutation array returned by `SortAllEdgesByDest`. (This could be a Python wrapper issue, but I doubt it).
* A crash when calling KTruss which seems to be *non-deterministic*.

Note: The Python tests are more aggressive than the C++ tests in some ways because they call the algorithms in more different ways and were written by me without knowing what the algorithm "expects" so they tend to exercise paths the original algorithm implementers didn't think of. So seeing a failure in the Python tests, but not C++ can still point to a C++ issue.

Ping me if you want to have a short meeting to talk about this. Co-debugging can be really helpful for this kind of thing and we can try to determine with certainty whether the bug is in the wrapper or in the C++ code.

arthurp reviewed (commented):
Request a rereview when it's time for me to be looking at this again.

arthurp commented on python/katana/_property_graph_numba.pyx.jinja:
```diff
@@ -14,10 +15,10 @@ from libc.stdint cimport uint32_t, uint64_t
 {% endcall %}
 {% call numba.method_with_body("edge_index", "uint64_t", ["uint64_t"]) %}
-    return self.topology().out_indices.get().Value(arg1)
+    return deref(self.topology().edge_end(arg1))
```
This needs to be `self.topology().edges(arg1).end()` I didn't expose `edge_end` in python because I hope it's going away.

arthurp commented on python/katana/_property_graph.pyx:
```diff
@@ -140,13 +142,13 @@ cdef class PropertyGraphBase:
         """
         cdef uint64_t prev
+        #TODO(amber): need to dereference GraphTopology::edge_begin()/edge_end()
+        # here
```
Why do you need to dereference them here? They are dereferenced below.

amberhassaan replied:
That was a TODO for you :). You already took care of it. I'll go ahead and remove it.

amberhassaan commented on python/katana/_property_graph.pyx:
```diff
@@ -140,13 +142,13 @@ cdef class PropertyGraphBase:
         Can be called from numba compiled code.
         """
         cdef uint64_t prev
+        #TODO(amber): need to dereference GraphTopology::edge_begin()/edge_end()
```
@arthurp : Can I remove the line above that declares a seemingly unused variable?

arthurp replied:
Yep. My bad.

amberhassaan commented:
> @amberhassaan I pushed updates to the python wrapper. The python tests are failing though. It's possible this is a bug in the wrapper, but there is also a crash in a call to `SortAllEdgesByDest` from the k-truss implementation, so I think the issues might be in the C++ implementation.
> 
> What I'm seeing are:
> 
> * Weird results of algorithms that don't match previous results. Note that the output checks are pretty strict, so if node order changed without actually affecting correctness that would explain some, _but not all_ of the failures.
> * Inconsistent output in the permutation array returned by `SortAllEdgesByDest`. (This could be a Python wrapper issue, but I doubt it).
> * A crash when calling KTruss which seems to be _non-deterministic_.
> 
> Note: The Python tests are more aggressive than the C++ tests in some ways because they call the algorithms in more different ways and were written by me without knowing what the algorithm "expects" so they tend to exercise paths the original algorithm implementers didn't think of. So seeing a failure in the Python tests, but not C++ can still point to a C++ issue.
> 
> Ping me if you want to have a short meeting to talk about this. Co-debugging can be really helpful for this kind of thing and we can try to determine with certainty whether the bug is in the wrapper or in the C++ code.

@arthurp : There was a bug I introduced in SortEdgesByDest() . I've fixed it since and this should resolve most if not all test failures. I'm surprised that CPP analytics tests, including triangle_counting, ktruss and everyone else using this function, didn't fail. We must have subpar verification?

roshandathathri commented:
> I'm surprised that CPP analytics tests, including triangle_counting, ktruss and everyone else using this function, didn't fail. We must have subpar verification?

We do sanity checks but not full output verification for shared-memory apps.

arthurp commented:
> We must have subpar verification?

Yes. Yes, we do.

amberhassaan commented:
> > I'm surprised that CPP analytics tests, including triangle_counting, ktruss and everyone else using this function, didn't fail. We must have subpar verification?
> 
> We do sanity checks but not full output verification for shared-memory apps.

My bug should have been insane for our sanity checks.

roshandathathri commented on libgalois/include/katana/ParallelSTL.h:
```diff
@@ -419,6 +419,27 @@ transform(
+      "Can't convert start_val to iterator's value_type");
+  static_assert(
+      std::is_arithmetic_v<T> && std::is_arithmetic_v<value_type>,
+      "iota only supported for numeric types");
```
Why are these asserts needed here? Shouldn't line 439 throw these asserts?

roshandathathri replied:
Sorry, I thought I submitted the above comment long ago. But apparently I hadn't!

amberhassaan replied:
I prefer keeping these checks as we're not sure how the standard library will respond. That's implementation dependent.

amberhassaan committed:
fix Python wrappers after name change. Missed by rebase

amberhassaan committed:
Cast to iterator's value_type

amberhassaan committed:
Need to pass sanitizer build. UndefinedBehaviorSanitizer doesn't like null pointer array access, i.e. &ptr[0] when ptr is NULL
